### PR TITLE
Map ArgumentSyntax nodes to appropriate IArgument operations

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/18405

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/analyzer-ioperation 

C# side is done.  Working on VB side now.

**Comment by CyrusNajmabadi** [comment]
@dotnet/analyzer-ioperation this is ready for first review pass.

**Comment by 333fred** [comment]
Is there any more generic way to accomplish this? We're going to have a number of these scenarios, and this method feels very fragile.

**Comment by CyrusNajmabadi** [comment]
> Is there any more generic way to accomplish this?

Not sure.  Do you have any ideas? :)

>We're going to have a number of these scenarios, and this method feels very fragile.

Can you clarify which additional scenarios you're thinking about.  I can try to test them.

**Comment by 333fred** [comment]
@CyrusNajmabadi pretty much all the bugs that jinu moved from me to you are this scenario. Off the top of my head, we have variable declarations in pretty much every scenario they can be used (fixed, for, using, multi-variable declarations) and case labels inside switch statements. I'll admit that I don't really have any good ideas on how to approach this problem, but at the very least lets not do a series of if/elses for each scenario.

**Comment by jinujoseph** [comment]
@CyrusNajmabadi , Fred is talking about all the issues listed here https://github.com/dotnet/roslyn/issues/18064

**Comment by CyrusNajmabadi** [comment]
> but at the very least lets not do a series of if/elses for each scenario.

I'm actually not certain why that's a bad approach.

**Comment by heejaechang** [comment]
I think @333fred want something more systematic or certain pattern. right now issue I believe is that, internally we have this indirection.

syntax node => bound node and bound node => operation

but we have desire to control what operation caller get on given syntax node. we could implement some kind of mechanism in operation which node should return this operation. or the reverse, each node explicitly specify what operation it will return.

what @CyrusNajmabadi did in this PR is basically that but @333fred might want it to be a bit more systematic like

if operation.sytnax  == given syntax then rethrn operation.

otherwise,

return operation.Search(given syntax)

so the logic is isolated to the related operation?

**Comment by CyrusNajmabadi** [comment]
I see.  I think there is merit in that suggestion.  However, i think what will likely happen is that as we add more cases we'll get a sense of what we would want to extract out.  Currently, i think this is acceptable as is, but may get refactored as we do more work.

**Comment by CyrusNajmabadi** [comment]
@dotnet/analyzer-ioperation please review.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs [hunk 1]
```diff
@@ -6,6 +6,7 @@
 using System.Linq;
 using Microsoft.CodeAnalysis.CSharp;
 using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs]: Why is this added?

File: src/Compilers/VisualBasic/Portable/Binding/MemberSemanticModel.vb [hunk 2]
```diff
@@ -794,6 +794,51 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         End Function
 
         Friend Overrides Function GetOperationWorker(node As VisualBasicSyntaxNode, options As GetOperationOptions, cancellationToken As CancellationToken) As IOperation
+            Dim argumentNode = TryCast(node, ArgumentSyntax)
+            If argumentNode IsNot Nothing Then
+                Return TryGetOperationForArgument(options, argumentNode, cancellationToken)
```
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Binding/MemberSemanticModel.vb]: Should we be returning Nothing in the event of failure to get the operation, or should we be returning GetBoundNodeSummaryAndCreateOperation?

File: src/Compilers/VisualBasic/Portable/Binding/MemberSemanticModel.vb [hunk 3]
```diff
@@ -794,6 +794,51 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+
+                Dim expression = TryCast(argumentNode, SimpleArgumentSyntax)?.Expression
+                If expression IsNot Nothing Then
+                    For Each argument In hasArguments.ArgumentsInEvaluationOrder
+                        If argument.Syntax Is expression Then
+                            Return argument
+                        End If
+
+                        If argument.ArgumentKind = ArgumentKind.ParamArray AndAlso
+                       TypeOf argument.Value Is IArrayCreationExpression Then
```
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Binding/MemberSemanticModel.vb]: Formatting

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs [hunk 4]
```diff
@@ -961,13 +961,46 @@ public override SyntaxTree SyntaxTree
...
+            var argumentExpression = argumentNode.Expression;
+            foreach (var argument in hasArguments.ArgumentsInEvaluationOrder)
+            {
+                if (argument.Syntax == argumentExpression)
+                {
+                    return argument;
+                }
+
+                if (argument.ArgumentKind == ArgumentKind.ParamArray &&
+                    argument.Value is IArrayCreationExpression arrayCreation)
```
Comment (genlu) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: I think the value can only be `IArrayCreationExpression` for `ParamArray`, we can move this out of the condition and make it an explicit conversion instead.

File: src/Compilers/VisualBasic/Portable/Binding/MemberSemanticModel.vb [hunk 5]
```diff
@@ -794,6 +794,51 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         End Function
 
         Friend Overrides Function GetOperationWorker(node As VisualBasicSyntaxNode, options As GetOperationOptions, cancellationToken As CancellationToken) As IOperation
+            Dim argumentNode = TryCast(node, ArgumentSyntax)
+            If argumentNode IsNot Nothing Then
```
Comment (genlu) [inline src/Compilers/VisualBasic/Portable/Binding/MemberSemanticModel.vb]: Why do we only check for argument syntax here but check both syntax and parent operation in C#? If call to `GetBoundNodeSummaryAndCreateOperation` inside `TryGetOperationForArgument` is not an `IHasArgumentsExpression`, `GetOperation` will return null, whereas in C#, it will return an operation from the bound node. Is this intended?

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs [hunk 6]
```diff
@@ -961,13 +961,46 @@ public override SyntaxTree SyntaxTree
...
+                    {
+                        if (arrayElement.Syntax == argumentExpression)
+                        {
+                            return argument;
+                        }
+                    }
+                }
+            }
+
+            return null;
```
Comment (genlu) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: Is there test covering this path?

File: src/Compilers/VisualBasic/Portable/Binding/MemberSemanticModel.vb [hunk 7]
```diff
@@ -794,6 +794,51 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+                                If child.Syntax Is expression Then
+                                    Return argument
+                                End If
+                            Next
+                        End If
+                    Next
+                End If
+            End If
+
+            Return Nothing
```
Comment (genlu) [inline src/Compilers/VisualBasic/Portable/Binding/MemberSemanticModel.vb]: Is there test covering this path?

File: src/Compilers/VisualBasic/Portable/Binding/MemberSemanticModel.vb (hunk 5 above)
Comment (CyrusNajmabadi) [inline src/Compilers/VisualBasic/Portable/Binding/MemberSemanticModel.vb]: Vb And C# behave differently here.  In C# asking for the BoundNode for an argument gives you the bound node from the containing invocation containing the argument.  However, in VB asking for the BoundNode for an argument just returns null.  We have to have different code in each language to deal with the discrepancies.

File: src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs (hunk 6 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: This is an tolerance path.  I'm uncertain on how i would hit it. However, i'm not comfortable with just throwing if this happens.   What is our IOperation policy here?

Comment (333fred) [inline src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs]: If it's a tolerance path, I think we should probably have a Debug assert, so if a test does hit it we'll know.

**Comment by jinujoseph** [comment]
@heejaechang  does getoperation work will address this and we are good to close this pr ?

**Comment by heejaechang** [comment]
Nope.


  *   heejae

**Comment by heejaechang** [comment]
@jinujoseph actually,  yes. confused this with other bug.

**Comment by heejaechang** [comment]
@jinujoseph we probably need to take tests in this PR.

**Comment by jinujoseph** [comment]
Thanks @heejaechang , Can you move the test over as separate PR pls
